File: backend/app/schemas/autodream.py
"""
AutoDream Pydantic Schemas
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


# ==================== Config Schemas ====================

class AutoDreamConfigResponse(BaseModel):
    id: int
    enabled: bool
    cron_expression: str
    batch_size: int
    max_consolidated_per_run: int
    archive_after_days: int
    delete_after_days: Optional[int] = None
    enable_dedup: bool
    enable_consolidation: bool
    enable_archival: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AutoDreamConfigUpdate(BaseModel):
    enabled: Optional[bool] = None
    cron_expression: Optional[str] = None
    batch_size: Optional[int] = Field(default=None, ge=1, le=500)
    max_consolidated_per_run: Optional[int] = Field(default=None, ge=1, le=100)
    archive_after_days: Optional[int] = Field(default=None, ge=1)
    delete_after_days: Optional[int] = Field(default=None, ge=1)
    enable_dedup: Optional[bool] = None
    enable_consolidation: Optional[bool] = None
    enable_archival: Optional[bool] = None


# ==================== Log Schemas ====================

class AutoDreamLogResponse(BaseModel):
    id: int
    triggered_at: datetime
    triggered_by: str
    status: str
    total_memories: int
    duplicates_found: int
    merged: int
    consolidated: int
    archived: int
    deleted: int
    llm_calls: int
    llm_tokens_in: int
    llm_tokens_out: int
    duration_seconds: float
    error_message: Optional[str] = None
    details: Optional[str] = None

    class Config:
        from_attributes = True


class AutoDreamLogListResponse(BaseModel):
    items: list[AutoDreamLogResponse]
    total: int


# ==================== Status / Trigger Schemas ====================

class AutoDreamStatusResponse(BaseModel):
    is_running: bool
    last_run: Optional[AutoDreamLogResponse] = None
    config: AutoDreamConfigResponse


class AutoDreamTriggerResponse(BaseModel):
    log_id: int
    status: str
    message: str

File: frontend/src/api/autodream.ts
import { api } from './index'

// ==================== Types ====================

export interface AutoDreamConfig {
  id: number
  enabled: boolean
  cron_expression: string
  batch_size: number
  max_consolidated_per_run: number
  archive_after_days: number
  delete_after_days: number | null
  enable_dedup: boolean
  enable_consolidation: boolean
  enable_archival: boolean
  created_at: string
  updated_at: string
}

export interface AutoDreamConfigUpdate {
  enabled?: boolean
  cron_expression?: string
  batch_size?: number
  max_consolidated_per_run?: number
  archive_after_days?: number
  delete_after_days?: number | null
  enable_dedup?: boolean
  enable_consolidation?: boolean
  enable_archival?: boolean
}

export interface AutoDreamLog {
  id: number
  triggered_at: string
  triggered_by: string
  status: string
  total_memories: number
  duplicates_found: number
  merged: number
  consolidated: number
  archived: number
  deleted: number
  llm_calls: number
  llm_tokens_in: number
  llm_tokens_out: number
  duration_seconds: number
  error_message: string | null
  details: string | null
}

export interface AutoDreamLogList {
  items: AutoDreamLog[]
  total: number
}

export interface AutoDreamStatus {
  is_running: boolean
  last_run: AutoDreamLog | null
  config: AutoDreamConfig
}

export interface AutoDreamTriggerResponse {
  log_id: number
  status: string
  message: string
}

// ==================== API ====================

export async function getAutoDreamStatus(): Promise<AutoDreamStatus> {
  const res = await api.get('/autodream/status')
  return res.data
}

export async function getAutoDreamConfig(): Promise<AutoDreamConfig> {
  const res = await api.get('/autodream/config')
  return res.data
}

export async function updateAutoDreamConfig(data: AutoDreamConfigUpdate): Promise<AutoDreamConfig> {
  const res = await api.put('/autodream/config', data)
  return res.data
}

export async function triggerAutoDream(): Promise<AutoDreamTriggerResponse> {
  const res = await api.post('/autodream/trigger')
  return res.data
}

export async function getAutoDreamLogs(page = 1, pageSize = 20): Promise<AutoDreamLogList> {
  const res = await api.get('/autodream/logs', {
    params: { page, page_size: pageSize },
  })
  return res.data
}

export async function getAutoDreamLogDetail(id: number): Promise<AutoDreamLog> {
  const res = await api.get(`/autodream/logs/${id}`)
  return res.data
}
